sdk/dotnet/Pulumi/X/Automation/LocalWorkspace.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Pulumi.X.Automation
+{
+    /// <summary>
+    /// LocalWorkspace is a default implementation of the Workspace interface.
+    /// <para/>
+    /// A Workspace is the execution context containing a single Pulumi project, a program,
+    /// and multiple stacks.Workspaces are used to manage the execution environment,
+    /// providing various utilities such as plugin installation, environment configuration
+    /// ($PULUMI_HOME), and creation, deletion, and listing of Stacks.
+    /// <para/>
+    /// LocalWorkspace relies on Pulumi.yaml and Pulumi.{stack}.yaml as the intermediate format
+    /// for Project and Stack settings.Modifying ProjectSettings will
+    /// alter the Workspace Pulumi.yaml file, and setting config on a Stack will modify the Pulumi.{stack}.yaml file.
+    /// This is identical to the behavior of Pulumi CLI driven workspaces.
+    /// </summary>
+    public class LocalWorkspace : IWorkspace
+    {
+        private readonly Task _readyTask;
+
+        public string WorkDir { get; }
+
+        public string? PulumiHome { get; }
+
+        public string? SecretsProvider { get; }
+
+        public PulumiFn? Program { get; set; }
+        
+        public IDictionary<string, string>? EnvironmentVariables { get; set; }
+
+        private LocalWorkspace(LocalWorkspaceOptions? options = null)
+        {
+            string? dir = null;
+            var readyTasks = new List<Task>();
+
+            if (options != null)
+            {
+                if (!string.IsNullOrWhiteSpace(options.WorkDir))
+                    dir = options.WorkDir;
+
+                this.PulumiHome = options.PulumiHome;
+                this.Program = options.Program;
+                this.SecretsProvider = options.SecretsProvider;
+
+                if (options.EnvironmentVariables != null)
+                    this.EnvironmentVariables = new Dictionary<string, string>(options.EnvironmentVariables);
+
+                if (options.ProjectSettings != null)
+                    readyTasks.Add(this.SaveProjectSettingsAsync(options.ProjectSettings));
+
+                if (options.StackSettings != null && options.StackSettings.Any())
+                {
+                    foreach (var pair in options.StackSettings)
+                        readyTasks.Add(this.SaveStackSettingsAsync(pair.Key, pair.Value));
+                }
+            }
+
+            if (string.IsNullOrWhiteSpace(dir))
+            {
+                dir = Path.Combine(Path.GetTempPath(), "automation-"); // TODO: should this not be randomized in some way?
```

Comment:
Should the `automation-` portion of this file path have some randomized part of it or do we intend to always be in the same directory?

Reply:
The API should support creating multiple of these "anonymous" working directories. For instance, two inline programs for project A and B could be created in separate working directories, but have stacks that are named the same and would collide if backed by the same workdir. In other words:

- projA + stackFoo = `pulumi.foo.yaml`
- projB + stackFoo = `pulumi.foo.yaml`

Reply:
Ok, that makes sense. So we should probably have `LocalWorkspace` implement `IDisposable` to cleanup that directory it created?

Reply:
This is a good idea. So far we've relied on the operating system to clean up any temp directories, but this would be good hygiene.